hazelcast-jet-core/src/main/java/com/hazelcast/jet/pipeline/SourceBuilder.java

```diff
@@ -0,0 +1,514 @@
+
+        private Stream() {
+        }
+
+        /**
+         * Sets the function that Jet will call whenever it needs more data from
+         * your source. The function receives the state object obtained from
+         * {@code createFn} and Jet's buffer object. It should add some items
+         * to the buffer, ideally those it can produce without making any blocking
+         * calls. On any given invocation the function may also choose not to add
+         * any items. Jet will automatically employ an exponential backoff strategy
+         * to avoid calling your function in a tight loop during a period when
+         * there's no data ready to be emitted.
+         *
+         * @param fillBufferFn function that fills the buffer with source data
+         * @param <T_NEW> type of the emitted items
+         * @return this builder with the item type reset to the one inferred from
+         *         {@code fillBufferFn}
+         */
+        @SuppressWarnings("unchecked")
+        public <T_NEW> Stream<T_NEW> fillBufferFn(
+                @Nonnull DistributedBiConsumer<? super S, ? super SourceBuffer<T_NEW>> fillBufferFn
+        ) {
+            Stream<T_NEW> newThis = (Stream<T_NEW>) this;
+            newThis.fillBufferFn = fillBufferFn;
+            return newThis;
+        }
+
+        /**
+         * Sets the function that Jet will call when cleaning up after a job has
+         * ended. It gives you the opportunity to release any resources held by the
+         * state object.
+         */
+        public Stream<T> destroyFn(@Nonnull DistributedConsumer<? super S> pDestroyFn) {
+            mDestroyFn = pDestroyFn;
+            return this;
+        }
+
+        /**
+         * Declares that you're creating a distributed source. On each member of
+         * the cluster Jet will create as many workers as you specify with the
+         * {@code preferredLocalParallelism} parameter. If you call this, you must
+         * ensure that all the source workers are coordinated and not emitting
+         * duplicated data. The {@code createFn} can consult {@link Processor.Context#totalParallelism()
+         * procContext.totalParallelism()} and {@link Processor.Context#globalProcessorIndex()
+         * procContext.globalProcessorIndex()}. Jet calls {@code createFn} exactly
+         * once with each {@code globalProcessorIndex} from 0 to {@code
+         * totalParallelism - 1}, this can help all the instances agree on which
+         * part of the data to emit.
+         *
+         * @param preferredLocalParallelism requested number of workers on each cluster member
+         */
+        public Stream<T> distributed(int preferredLocalParallelism) {
+            mPreferredLocalParallelism = preferredLocalParallelism;
+            return this;
+        }
+
+        /**
+         * Builds and returns the unbounded stream source.
+         */
+        public StreamSource<T> build() {
```

Comment:
missing `@Nonnull`. Same in TimestampedStreamSource, so omitting those comments.